ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,938 @@ PHP_FUNCTION(bcscale)
+		op1 = &op1_copy;
+	}
+
+	bc_num ret = NULL;
+	size_t scale;
+	switch (opcode) {
+		case ZEND_ADD:
+			bcmath_number_add_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true);
+			break;
+		case ZEND_SUB:
+			bcmath_number_sub_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true);
+			break;
+		case ZEND_MUL:
+			bcmath_number_mul_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true);
+			break;
+		case ZEND_DIV:
+			if (UNEXPECTED(bcmath_number_div_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true) == FAILURE)) {
+				goto fail;
+			}
+			break;
+		case ZEND_MOD:
+			if (UNEXPECTED(bcmath_number_mod_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true) == FAILURE)) {
+				goto fail;
+			}
+			break;
+		case ZEND_POW:
+			if (UNEXPECTED(bcmath_number_pow_internal(n1, n2, &ret, n1_full_scale, n2_full_scale, &scale, true, true) == FAILURE)) {
+				goto fail;
+			}
+			break;
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
+
+	if (Z_TYPE_P(op1) != IS_OBJECT) {
+		bc_free_num(&n1);
+	}
+	if (Z_TYPE_P(op2) != IS_OBJECT) {
+		bc_free_num(&n2);
+	}
+
+	bcmath_number_obj_t *intern = bcmath_number_new_obj(ret, scale);
+	ZVAL_OBJ(ret_val, &intern->std);
+
+	/* For increment and decrement, etc */
+	if (op1 == &op1_copy) {
+		zval_ptr_dtor(op1);
+	}
+
+	return SUCCESS;
+
+fail:
+	if (Z_TYPE_P(op1) != IS_OBJECT) {
+		bc_free_num(&n1);
+	}
+	if (Z_TYPE_P(op2) != IS_OBJECT) {
+		bc_free_num(&n2);
+	}
+	return FAILURE;
+}
+
+static int bcmath_number_compare(zval *op1, zval *op2)
+{
+	zend_object *obj1 = NULL;
+	zend_string *str1 = NULL;
+	zend_long lval1;
+
+	zend_object *obj2 = NULL;
+	zend_string *str2 = NULL;
+	zend_long lval2;
+
+	bc_num n1 = NULL;
+	bc_num n2 = NULL;
+
+	if (UNEXPECTED(Z_TYPE_P(op1) == IS_NULL)) {
+		lval1 = 0;
+	} else if (UNEXPECTED(bcmath_number_parse_num(op1, &obj1, &str1, &lval1) == FAILURE)) {
+		goto fallback;
+	}
+
+	if (UNEXPECTED(Z_TYPE_P(op2) == IS_NULL)) {
+		lval2 = 0;
+	} else if (UNEXPECTED(bcmath_number_parse_num(op2, &obj2, &str2, &lval2) == FAILURE)) {
+		goto fallback;
+	}
+
+	if (UNEXPECTED(bc_num_from_obj_or_str_or_long(&n1, NULL, obj1, str1, lval1) == FAILURE ||
+		bc_num_from_obj_or_str_or_long(&n2, NULL, obj2, str2, lval2) == FAILURE)) {
+		goto fallback;
+	}
+
+	int ret = bc_compare(n1, n2, MAX(n1->n_scale, n2->n_scale));
+
+	if (Z_TYPE_P(op1) != IS_OBJECT) {
+		bc_free_num(&n1);
+	}
+	if (Z_TYPE_P(op2) != IS_OBJECT) {
+		bc_free_num(&n2);
+	}
+
+	return ret;
+
+fallback:
+	return zend_std_compare_objects(op1, op2);
+}
+
+#define BCMATH_NUMBER_METHOD_PARSE_NUM(zv, obj, str, lval, arg_num) do { \
+	if (UNEXPECTED(bcmath_number_parse_num(zv, &obj, &str, &lval) == FAILURE)) { \
+		zend_argument_type_error(arg_num, "must be of type int, string, or %s, %s given", ZSTR_VAL(bcmath_number_ce->name), zend_zval_value_name(zv)); \
+		RETURN_THROWS(); \
+	} \
+} while (0)
+
+#define BCMATH_NUMBER_METHOD_PARSE_SCALE(zv, lval, is_null, arg_num) do { \
```

Comment:
There are currently no macros that can parse int|string|Number, so for that we have to get the ZVAL and parse it ourselves.

If use the standard macro for scale, the parsing order of the arguments will be messed up, and an error for the second argument will be output before the first argument.

Reply:
I see. I think you can introduce a variant of Z_PARAM_ macros for these (in the bcmath.c file) so that the argument parsing becomes more like how it's done in the standard way.

Reply:
Added a comment in https://github.com/php/php-src/pull/13741/commits/6c6d0d50363b08b320b72123c12e61a405515965

Reply:
Fixed in https://github.com/php/php-src/pull/13741/commits/0fbc1776c14c7667400bfc825ed38468e67862bb